Allow setting 'rel' and 'title' attributes for stylesheets.

Jellby opened the pull request:
This addresses #1764. I added documentation too, but no test (I'm not sure how that works). Please tell me if there is something else wrong or missing.

Commit:
Allow setting 'rel' and 'title' attributes for stylesheets.

Commit:
Fix two statements in one line

lehmannro commented on sphinx/application.py:
```diff
@@ -681,14 +681,19 @@ def add_javascript(self, filename):
                 posixpath.join('_static', filename))
 
-    def add_stylesheet(self, filename):
+    def add_stylesheet(self, filename, rel=None, title=None):
```
I don't think passing arbitrary strings to `rel` is good API design.  This should be some kind of enum or, which I think would be totally sufficient for the use cases at hand, just an `alternate=bool` flag.

Jellby replied:
Better with the new commit?

Commit:
Changed 'rel' to 'alternate'

Commit:
Merge remote-tracking branch 'upstream/master'

Jellby commented:
How can I create a test? The [Developer's Guide](http://sphinx-doc.org/latest/devguide.html) says "Write a test", but how do I write it and how do I specify what the correct result is? In this case I need to test that the add_stylesheet method accepts the two additional arguments and that setting the css_files in templates works as intended (i.e., check the html builder with custom conf.py and modified template).

EDIT: Nevermind, I think I figured it out (see next commit).

Commit:
Test for alternate stylesheets

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

shimizukawa commented on sphinx/themes/basic/layout.html:
```diff
@@ -103,7 +103,25 @@
+                {%- set alt = False %}
+            {%- endif %}
+    <link rel="{% if alt %}alternate {% endif %}stylesheet" href="{{ pathto(filename, 1) }}" type="text/css" />
+        {%- endif %}
```
This logic shouldn't be wrote on html template because it makes difficult to read and/or understand for developers and users. LGTM except this lines.

Jellby replied:
But it has to be in the template if I want to allow css_files being modified in a template file (see tests/roots/test-stylesheets/_templates/layout.html). Or can I process css_files between its definition with:

```
{% set css_files = ... %}
```

and its implicit use with:

```
{{ css() }}
```

?

The problem is that the default value for "alternate", if it's not specified, depends on whether or not a "title" attribute was provided. I can easily handle this on the python side (in add_stylesheet), but that's not the only way to add css files.

shimizukawa replied:
IMO, css_files variable must not have multiple structure according to circumstances because it is difficult for sphinx users to understand and to tweak template. And also it is hard for maintainers to keep it stable.

BTW, I reconsidered about this change. Then I can't merge this for 1.4 release because side effect of this change is large for users who are using custom template that references to `css_files`. To resolve this concern, we should provide another variable.

Commit:
Merge remote-tracking branch 'upstream/master'

Jellby committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	tests/test_build_html.py

Commit:
Merge remote-tracking branch 'upstream/master'

Jellby committed:
Merge remote-tracking branch 'upstream/master'

Jellby committed:
Move CSS properties to a different variable

Jellby commented:
@shimizukawa I modified it to use a separate variable for the properties. This should make it completely backwards-compatible. The logic in the template is also slightly clearer now, I believe, although I don't like the way to update a dict in jinja.

Jellby committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	tests/test_build_html.py

Jellby committed:
Remove trailing whitespaces

Jellby committed:
Let's see now

shimizukawa commented on sphinx/themes/basic/layout.html:
```diff
@@ -103,8 +103,24 @@
+        {%- else %}
+            {%- set rel = "stylesheet" %}
+        {%- endif %}
+    <link rel="{{ rel }}" href="{{ pathto(cssfile, 1) }}" type="text/css"{% if csstitle is defined %} title="{{ csstitle }}"{% endif %} />
```
`css_props` is nice, however this if/else/set structure is too complicated to read.
My idea I mentioned is like this:

```
    {#- Process the custom css files, with their alternate and title properties if available #}
    {%- for cssfile in css_files %}
    <link rel="{{ css_props[cssfile].rel }}" href="{{ pathto(cssfile, 1) }}" type="text/css" title="{{ css_props[cssfile].csstitle }}" />
    {%- endfor %}
```

Trade-off is, `title=""` will be rendered if title is not provided.
@Jellby how about this?

Jellby replied:
I'm afraid that won't work. In HTML/CSS a stylesheet with a `title` attribute is treated differently from another without it. A stylesheet without `title` is always applied (persistent). A stylesheet with `title` is applied by default, but is disabled if another alternate stylesheet is selected (preferred). See https://developer.mozilla.org/en/docs/Correctly_Using_Titles_With_External_Stylesheets for example. Your suggestion would make it impossible to specify persistent stylesheets (unless a browser considers an empty title as no title, but I haven't found anything in the specification indicating this should be the case).

And then, I would like to allow setting the properties in a template file, not only through the `add_stylesheet` method, I believe it can be useful or needed in some cases, probably when designing a theme. If the code is difficult to read, would it help if I add some comments explaining its purpose?

tk0miya replied:
Agreed, `title=""` is not good. But I also agreed to this is too complicated.

I prefer to pack the CSS information to single object in python side.
For example:
```
{%- for css in css_files %}
<link rel="{{ css.rel }}" href="{{ pathto(css.filename, 1) }}" type="text/css"{% if css.title is none %} title="{{ css.title }}"{% endif %} />
<%- endfor %}
```

How about this?

shimizukawa commented on tests/roots/test-stylesheets/_templates/layout.html:
```diff
@@ -0,0 +1,7 @@
+    {"_static/more_default.css": {"title": "Default", "alternate": False},
+     "_static/more_alternate1.css": {"title": "Alternate"},
+     "_static/more_alternate2.css": {"alternate": True} }
+) %}
```
I don't think the functionality is needed to modify title and alternate attribute from template side.

Jellby replied:
As pointed out above, what if I want to set that in a theme? Can a theme use python code?

shimizukawa commented on sphinx/application.py:
```diff
@@ -796,14 +796,20 @@ def add_javascript(self, filename):
                 posixpath.join('_static', filename))
 
-    def add_stylesheet(self, filename):
+    def add_stylesheet(self, filename, alternate=None, title=None):
```
default value of alternate must be `False`

Jellby replied:
That could be. Personally, I'd say it is more often that one wants to set a stylesheet with title as alternate, so I wanted to make the "default" True if there is a title and False if there isn't. To get this I had to store somehow whether the alternate argument had been given or not: basically, to pass the unprocessed information to the template (see above).

shimizukawa commented on sphinx/application.py:
```diff
@@ -796,14 +796,20 @@ def add_javascript(self, filename):
+        if alternate is not None:
+            props['alternate'] = bool(alternate)
+        if title is not None:
+            props['title'] = title
```
Sometime 'alternate' and 'title' keys are given, sometime not.
I think `props = {'alternate': alternate, 'title': title}` is simple and better.

Jellby replied:
If I do that, and one doesn't pass a title, would e.g. `{% if css_props[cssfile].title is defined %}` work in the template? From a python point of view, it is defined (as `None`).

shimizukawa commented on doc/templating.rst:
```diff
@@ -202,8 +202,24 @@ Overriding works like this::
-   Similar to :data:`script_files`, for CSS files.
+   Similar to :data:`script_files`, for CSS files::
 
+      {% set css_files = css_files + ["_static/mystyle.css"] %}
```
IMO, rewriting a template variable is bad manner.
Even if you or someone want to use such workaround, official document should not provide such information.

Jellby replied:
But it's already there for `script_files`...

tk0miya replied:
+1 for shimizukawa.
I feel this is hacky approach too. So it should not be described in our document.
I think our templates around `script_files` should be also fixed.

To do that, it would be nice to add new extensible block like following:
```
{%- block css %}
  {{ super() }}
  <link rel="stylesheet" href="_static/mystyle.css" type="text/css" />
{%- endblock %}
```

shimizukawa commented on doc/templating.rst:
```diff
@@ -202,8 +202,24 @@ Overriding works like this::
+   attributes for the css files. The keys are the css filenames, the values
+   are dicts themselves:
+
+      {% set _dummy = css_props.update( {"_static/mystyle.css": {"alternate":False, "title":"Default"} }) %}
```
same as above.

However, `css_props` data structure information is useful for everyone.
Can you update please?

shimizukawa reviewed (changes requested):
Reviewed. @Jellby please take a look.

Jellby committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	sphinx/application.py
	sphinx/builders/html.py

Jellby committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	tests/test_build_html.py

Jellby committed:
Update version (there's still hope)

Jellby committed:
Try to correct syntax for type annotation

Jellby committed:
again

Jellby committed:
again

Jellby committed:
again

Jellby committed:
again

Jellby commented:
@shimizukawa Would you please reconsider your objections, or reply? The HTML logic for alternate stylesheets is a bit involved, and I really want to set alternate stylesheets in a template file, as part of a theme.

tk0miya commented:
+1 for basic idea, but we need to arrange the code.

Jellby committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	tests/test_build_html.py

Jellby committed:
Rewrite and simplify stylesheet handling

Thanks to tk0miya's comment, I learnt one can add stuff to template
blocks, that allows a much simpler stylesheet configuration, considering
that changes at the template level will be more... well, low-level.
Hopefully this is now acceptable.

Jellby commented:
@shimizukawa @tk0miya That suggestion about the template block and the reference to `super()` solved all my problems. I can now get rid of all the complicate template logic and assume that if anyone (like myself) wants to modify the template, they can write HTML code directly. After that, I could also implement the other suggestions/requests. I added a ``scripts`` block too, I figured that doesn't harm.

tk0miya commented:
LGTM! Thank you for update :-)

tk0miya commented:
ping @shimizukawa 
Could you review this please? It's time to v1.6!

shimizukawa commented on sphinx/themes/basic/layout.html:
```diff
@@ -105,8 +105,8 @@
-    {%- for cssfile in css_files %}
-    <link rel="stylesheet" href="{{ pathto(cssfile, 1) }}" type="text/css" />
+    {%- for css in css_files %}
+    <link rel="{{ css.rel }}" href="{{ pathto(css.filename, 1) }}" type="text/css"{% if css.title is not none %} title="{{ css.title }}"{% endif %} />
```
This is a incompatible change.
I think it would be better if `css` object (current implementation is using dict) has `__str__` and return a value of `filename` attribute to keep compatibility.
@tk0miya How do you think?

tk0miya replied:
Yes, you're right. I'll arrange that after merging.

shimizukawa reviewed (approved):
Reviewed. Thanks @Jellby!

@tk0miya I leave a comment for about incompatibility.

tk0miya commented:
@Jellby Thank you for long-term and tough contributing! I'm glad to merge this :-)
